mcmc_utils.py

```diff
@@ -280,9 +286,49 @@ def flatchain(chain, npars, nskip=0, thin=1):
     the steps of the chain are highly correlated'''
     return chain[:, nskip::thin, :].reshape((-1, npars))
 
-
-def readchain(file, nskip=0, thin=1):
-    data = pd.read_csv(file, header=None, compression=None, delim_whitespace=True)
+def reverse_readline(filename, buf_size=8192):
+    """A generator that returns the lines of a file in reverse order
+    Taken from
+    https://stackoverflow.com/questions/2301789/read-a-file-in-reverse-order-using-python
+    """
+    with open(filename) as fh:
+        segment = None
+        offset = 0
+        fh.seek(0, os.SEEK_END)
+        file_size = remaining_size = fh.tell()
+        while remaining_size > 0:
+            offset = min(file_size, offset + buf_size)
+            fh.seek(file_size - offset)
+            buffer = fh.read(min(remaining_size, buf_size))
+            remaining_size -= buf_size
+            lines = buffer.split('\n')
+            # the first line of the buffer is probably not a complete line so
+            # we'll save it and append it to the last line of the next buffer
+            # we read
+            if segment is not None:
+                # if the previous chunk starts right from the beginning of line
+                # do not concact the segment to the last line of new chunk
+                # instead, yield the segment first
+                if buffer[-1] != '\n':
+                    lines[-1] += segment
+                else:
+                    yield segment
+            segment = lines[0]
+            for index in range(len(lines) - 1, 0, -1):
+                if lines[index]:
+                    yield lines[index]
+        # Don't yield None if the file was empty
+        if segment is not None:
+            yield segment
+
+def readchain(file, columns=None, headers=None):
```

Comment:
I guess this change is to allow you to only read certain columns into memory? Thats a nice idea, but it's not so obvious that's what's happening, and you should add a docstring. I think the code could be more elegant as 

```python
kwargs = {'compression': None, 'delim_whitespace': True}
if columns is not None:
    kwargs['usecols'] = columns
data = pd.read_csv(file, **kwargs)
``` 

So you might consider that.

Finally, what is the ```headers``` argument used for, if anything?

Reply:
Yes, that was the idea with `columns`, I added it because pandas uses less memory with that flag, so I could read in chains that were bigger than my memory. You're right though, I definitely need a docstring. I'll add this.